### ISSUE TITLE
Unknown argument type with itemgetter

### ISSUE THREAD

**Issue by mehdigmira** [microsoft/pyright#5693]
**Describe the bug**
In some specific conditions itemgetter is inferred as itemgetter[Unknown], starting from 1.1.320

**Code or Screenshots**

```python
# pyright: strict


from collections.abc import Callable
from operator import itemgetter
from typing import Any


def merge(
    func1: Callable[..., Any],
    func2: Callable[..., Any] = ...,
) -> Any:
    ...


# this causes
# error: Argument type is partially unknown
#     Argument corresponds to parameter "func2" in function "merge"
#     Argument type is "itemgetter[Unknown]" (reportUnknownArgumentType)
merge(itemgetter("foo"), itemgetter("bar"))

# but this reports no error
x = itemgetter("bar")
merge(itemgetter("foo"), x)
```

**Comment by erictraut** [comment]
The type declaration for `itemgetter` (in typeshed's `_operator.pyi`) contains some questionable overloads for `__new__`. I'll copy two such overloads to demonstrate what I mean.

```python
class itemgetter(Generic[_T_co]):
    @overload
    def __new__(cls, item: _T_co) -> itemgetter[_T_co]: ...
    @overload
    def __new__(cls, item: _T_co, __item2: _T_co) -> itemgetter[tuple[_T_co, _T_co]]: ...
```

The problem here is that `__new__` for type `X[T]` should return type `X[T]`, not construct some other type. This becomes important if you explicitly specialize the type before calling the constructor. For example, what happens if you call `itemgetter[str]("", "")`? You are expecting to instantiate an `itemgetter[str]`, but the `__new__` method constructs an `itemgetter[tuple[str, str]]` instead.

I fixed another bug relating to a similar issue in 1.1.320, so that's probably responsible for the change in behavior.

I'll see if there's something I can do to improve the behavior in this case and avoid the `reportUnknownArgumentType` diagnostic.

**Comment by erictraut** [comment]
The problem turned out to be unrelated to the way `itemgetter.__new__` was defined. The problem would have occurred for any generic class that was passed as an argument more than once to the same call. 

This will be fixed in the next release.

**Comment by mehdigmira** [comment]
Does it also fix the following snippet ?

```python
from __future__ import annotations

from typing import *

_T = TypeVar("_T")


class ORMDescriptor(Generic[_T]):
    @overload
    def __get__(self, instance: Any, owner: Literal[None]) -> ORMDescriptor[_T]:
        ...

    @overload
    def __get__(self, instance: Literal[None], owner: Any) -> Any:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T:
        ...

    def __get__(self, instance: object, owner: Any) -> Union[ORMDescriptor[_T], Any, _T]:
        ...


class A:
    y: ORMDescriptor[str]

    @property
    def label_descriptions(self):
        reveal_type(
            self.y
        )  # information: Type of "self.y" is "Unknown", but if i remove the first overload of __get__ in ORMDescriptor, it is "str"
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.324, which I just published. It will also be included in a future release of pylance.
